[RNMobile] Initial HTML E2E test

jd-alexander opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->
Fixes https://github.com/WordPress/gutenberg/issues/26146

Gutenberg Mobile PR https://github.com/wordpress-mobile/gutenberg-mobile/pull/2753

## Description
This change introduces a React Native E2E test that renders the initial HTML content that's used during development. 


## How has this been tested?
I tested this on Android, by running `TEST_RN_PLATFORM=android npm run native device-tests:local gutenberg-editor-blocks.test`. I will await the results of this PR for iOS because I have been having some issues when running iOS tests locally.

## Types of changes
- The test that's run on `gutenberg` PR has been changed from the `Gallery` test to the `initial html` test. 
- A function was created to scroll to the last block within the post to ensure that all blocks are rendered. I noticed that the test took a while to run because the `swipe` functionality within the test helper waits 3 seconds each time a swipe is done. I am assuming this is done to ensure that all elements on the screen are loaded and such forth so I didn't modify it. I am sharing this here in case this needs to be optimized.
- the `initial html` content was added to the `test-data.js` file. I am not sure if there's a way to include the `initial html` being used in the other part of the app. If there is, please let me know so I can reduce the duplicate code. I just utilized this approach since I saw that all the test data resided in the test directory and I wasn't able to import the `initialHtml` module. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

jd-alexander committed:
added initial-html to test-data.js

jd-alexander committed:
Added a test that loads all the blocks within a post.

jd-alexander committed:
Updated the ci test runners to run gutenberg-editor-blocks

jd-alexander committed:
added a function that scrolls to a specific element and returns it.

jd-alexander committed:
updated the test to utilize the scroll and return functionality

jd-alexander commented:
So as seen above the test was run successfully. I did download the video generated from Saucelabs and I noticed that the Android test video length was 3mins and the iOS test video length was 5 mins. Just noting that here for discussion :)

ceyhun commented:
Thanks a lot for this @jd-alexander 🎉 

>I will await the results of this PR for iOS because I have been having some issues when running iOS tests locally.

There could be some issues if you're using Xcode 12: https://github.com/WordPress/gutenberg/pull/26283#pullrequestreview-514303259. You can install Xcode 11 aside 12 for now until that PR is merged.

> the initial html content was added to the test-data.js file. I am not sure if there's a way to include the initial html being used in the other part of the app. If there is, please let me know so I can reduce the duplicate code. I just utilized this approach since I saw that all the test data resided in the test directory and I wasn't able to import the initialHtml module.

I think `import initialHtml from '../src/initial-html';` should work in this case.

>So as seen above the test was run successfully. I did download the video generated from Saucelabs and I noticed that the Android test video length was 3mins and the iOS test video length was 5 mins. Just noting that here for discussion :)

I wasn't able to find the Sauce Labs video link for this test in the CircleCI output but I was able to watch it from GitHub Actions artifacts.

For iOS I checked this run: https://github.com/WordPress/gutenberg/pull/26405/checks?check_run_id=1294846308. The most time consuming thing was typing the content in the text view. It took 5 minutes to type the whole initialHtml. I think a way to paste the initialHtml instead of typing on iOS would be really nice. It also didn't seem to scroll the list at all at the end of the test (maybe just a few pixels).

For Android I checked the artifacts in this run: https://github.com/WordPress/gutenberg/pull/26405/checks?check_run_id=1294846311. It was very fast to type the content (maybe it was pasting) and the rest of the time (~2:30mins) was taken by scrolling. It was waiting 3 seconds each time like you mentioned but the scrolling itself seemed very slow as well.

jd-alexander commented:
> Thanks a lot for this @jd-alexander 🎉

Thanks for your review @ceyhun 🙇 

> There could be some issues if you're using Xcode 12: #26283 (review). You can install Xcode 11 aside 12 for now until that PR is merged.

Thanks I am going to do that. 

> I think import initialHtml from '../src/initial-html'; should work in this case.

Nice 😄 

> For iOS I checked this run: https://github.com/WordPress/gutenberg/pull/26405/checks?check_run_id=1294846308. The most time consuming thing was typing the content in the text view. It took 5 minutes to type the whole initialHtml. I think a way to paste the initialHtml instead of typing on iOS would be really nice. It also didn't seem to scroll the list at all at the end of the test (maybe just a few pixels).

Indeed! I am going to work on a way of doing the typing more instantaneously and yes, I am also going to look into why it doesn't do the scroll completely. I am wondering if it because it discover the element so it just stops. I will see what I can come up with. 

> For Android I checked the artifacts in this run: https://github.com/WordPress/gutenberg/pull/26405/checks?check_run_id=1294846311. It was very fast to type the content (maybe it was pasting) and the rest of the time (~2:30mins) was taken by scrolling. It was waiting 3 seconds each time like you mentioned but the scrolling itself seemed very slow as well

Indeed. I am going to modify the function that does the swipes so that I can pass in a delay param so that the scrolling doesn't operate so slowly.

jd-alexander committed:
utilize a shared initialHtml for both test and initial editor content.

jd-alexander committed:
added delay to swipe methods.

jd-alexander commented:
So today I did some work to make these changes more optimized. 

- [x] I got the swipe functionality to delay 
- [x] I also utilized the `initialHtml` from the `src` and removed it from the `testData`
------------------

From an iOS standpoint, I still didn't get the test running on iOS 😞  

This is the output I am getting 

```
   [init({"browserName":"","platformName":"iOS","platformVersion":"13.4","deviceName":"iPhone 11","os":"iOS","deviceOrientation":"portrait","automationName":"XCUITest","appiumVersion":"1.16.0","app":"/Users/joeldean/Projects/gutenberg/packages/react-native-editor/ios/build/GutenbergDemo/Build/Products/Release-iphonesimulator/GutenbergDemo.app","processArguments":{"args":["uitesting"]}})] The environment you requested was unavailable.
    An unknown server-side error occurred while processing the command. Original error: Bad app: /Users/joeldean/Projects/gutenberg/packages/react-native-editor/ios/build/GutenbergDemo/Build/Products/Release-iphonesimulator/GutenbergDemo.app. App paths need to be absolute or an URL to a compressed app file: The application at '/Users/joeldean/Projects/gutenberg/packages/react-native-editor/ios/build/GutenbergDemo/Build/Products/Release-iphonesimulator/GutenbergDemo.app' does not exist or is not accessible
```

I also tried to utilize the copy and paste functionality, within the app using this function I wrote below. The weird thing that I think is happening is the content being escaped by the driver.

```
	// copy and paste html editor content explicitly
	async copyAndPasteHtmlContent( html ) {
		await toggleHtmlMode( this.driver, true );

		await this.driver.setClipboard(html,'plaintext');
		const htmlContentView = await this.getTextViewForHtmlViewContent();

		await longPressMiddleOfElement( this.driver, htmlContentView );
		await tapPasteAboveElement( this.driver, htmlContentView );

		await toggleHtmlMode( this.driver, false );
	}
```

The escaped content looks like 

```
<!-- wp:heading -->\n<h2>Text Blocks</h2>\n<!-- /wp:heading -->\n\n<!-- wp:heading -->\n<h2 id=\"this-is-an-anchor\">What is Gutenberg?</h2>\n<!-- /wp:heading -->\n\n<!-- wp:paragraph -->\n<p><strong>Bold</strong> <em>Italic</em> <s>Striked</s> Superscript<sup>(1)</sup> Subscript<sub>(2)</sub> <a href=\"http://www.wordpress.org\" target=\"_blank\" rel=\"noreferrer noopener\">Link</a></p>\n<!-- /wp:paragraph -->\n\n<!-- wp:heading {\"level\":4} -->\n<h4>List</h4>\n<!-- /wp:heading -->\n\n<!-- wp:list -->\n<ul><li>First Item</li><li>Second Item</li><li>Third Item</li></ul>\n<!-- /wp:list -->\n\n<!-- wp:heading {\"level\":4} -->\n<h4>Quote</h4>\n<!-- /wp:heading -->\n\n<!-- wp:quote {\"align\":\"left\",\"className\":\"is-style-large\"} -->\n<blockquote class=\"wp-block-quote has-text-align-left is-style-large\"><p>\"This will make running your own blog a viable alternative again.\"</p><cite>— <a href=\"https://twitter.com/azumbrunnen_/status/1019347243084800005\">Adrian Zumbrunnen</a></cite></blockquote>\n<!-- /wp:quote -->\n\n<!-- wp:pullquote -->\n<figure class=\"wp-block-pullquote\"><blockquote><p>One of the hardest things to do in technology is disrupt yourself.</p><cite>Matt Mullenweg</cite></blockquote></figure>\n<!-- /wp:pullquote -->\n\n<!-- wp:heading {\"level\":4} -->\n<h4>Style Paragraph</h4>\n<!-- /wp:heading -->\n\n<!-- wp:paragraph {\"dropCap\":true,
```

I haven't figured out what exactly is causing this as yet. 

Also while I was watching the video, I think the code that presses the paste button may be outdated so I will need to run a check on the coordinate algorithm being used to determine where on the screen should be clicked.

ceyhun commented:
> From an iOS standpoint, I still didn't get the test running on iOS 😞

Is the folder `gutenberg/packages/react-native-editor/ios/build/GutenbergDemo/Build/Products/Release-iphonesimulator/GutenbergDemo.app` empty in this case? 

Are you using Xcode 11 on the command-line? You can check via running `xcodebuild -version` and switch using `sudo xcode-select --switch /Applications/Xcode_11_path.app`

jd-alexander commented:
Thanks much @ceyhun that fixed the XCode issue. I am now having another issue 

```
    [init({"browserName":"","platformName":"iOS","platformVersion":"13.4","deviceName":"iPhone 11","os":"iOS","deviceOrientation":"portrait","automationName":"XCUITest","appiumVersion":"1.16.0","app":"/Users/joeldean/Projects/gutenberg/packages/react-native-editor/ios/build/GutenbergDemo/Build/Products/Release-iphonesimulator/GutenbergDemo.app","processArguments":{"args":["uitesting"]}})] The environment you requested was unavailable.
    An unknown server-side error occurred while processing the command. Original error: Bad app: /Users/joeldean/Projects/gutenberg/packages/react-native-editor/ios/build/GutenbergDemo/Build/Products/Release-iphonesimulator/GutenbergDemo.app. App paths need to be absolute or an URL to a compressed app file: The application at '/Users/joeldean/Projects/gutenberg/packages/react-native-editor/ios/build/GutenbergDemo/Build/Products/Release-iphonesimulator/GutenbergDemo.app' does not exist or is not accessible
```

So it seems the GutenbergDemo release project is not being generated 

> Is the folder gutenberg/packages/react-native-editor/ios/build/GutenbergDemo/Build/Products/Release-iphonesimulator/GutenbergDemo.app empty in this case

Yes, you are indeed correct with this one. 

Following the testing steps in the README.md file I ran the appium doctor commands for ios and I got the output below, which means I need to fix some issues ( some of them seem unrelated) so I will do that to see if it will help. 

```
info AppiumDoctor ### Diagnostic for optional dependencies starting ###
WARN AppiumDoctor  ✖ opencv4nodejs cannot be found.
WARN AppiumDoctor  ✖ ffmpeg cannot be found
WARN AppiumDoctor  ✖ mjpeg-consumer cannot be found.
WARN AppiumDoctor  ✖ set-simulator-location is not installed
WARN AppiumDoctor  ✖ idb and idb_companion are not installed
WARN AppiumDoctor  ✖ applesimutils cannot be found
WARN AppiumDoctor  ✖ ios-deploy cannot be found
info AppiumDoctor ### Diagnostic for optional dependencies completed, 7 fixes possible. ###
info AppiumDoctor
info AppiumDoctor ### Manual Fixes Needed ###
info AppiumDoctor The configuration cannot be automatically fixed, please do the following first:
WARN AppiumDoctor  ➜ Please install Carthage. Visit https://github.com/Carthage/Carthage#installing-carthage for more information.
info AppiumDoctor
info AppiumDoctor ### Optional Manual Fixes ###
info AppiumDoctor The configuration can install optionally. Please do the following manually:
WARN AppiumDoctor  ➜ Why opencv4nodejs is needed and how to install it: https://github.com/appium/appium/blob/master/docs/en/writing-running-appium/image-comparison.md
WARN AppiumDoctor  ➜ ffmpeg is needed to record screen features. Please read https://www.ffmpeg.org/ to install it
WARN AppiumDoctor  ➜ mjpeg-consumer module is required to use MJPEG-over-HTTP features. Please install it with 'npm i -g mjpeg-consumer'.
WARN AppiumDoctor  ➜ set-simulator-location is needed to set location for Simulator. Please read https://github.com/lyft/set-simulator-location to install it
WARN AppiumDoctor  ➜ Why idb is needed and how to install it: https://github.com/appium/appium-idb
WARN AppiumDoctor  ➜ Why applesimutils is needed and how to install it: http://appium.io/docs/en/drivers/ios-xcuitest/
WARN AppiumDoctor  ➜ ios-deploy is used as a fallback command to install iOS applications to real device. Please read https://github.com/ios-control/ios-deploy/ to install it
info AppiumDoctor
info AppiumDoctor ###
info AppiumDoctor
info AppiumDoctor Bye! Run appium-doctor again when all manual fixes have been applied!
```

jd-alexander commented:
@ceyhun would you mind giving me some help on the iOS side of things?[ I lessened the swipe delay and other updates ](https://github.com/WordPress/gutenberg/pull/26405#issuecomment-716894328)and that is working great but on iOS, I still haven't figured out how to get the pasting working right. I wanted to dig a bit deeper today but I am wrangling the release process for the first time so it ending up consuming most of my time. So yes, let me know if you are able to have a look as I want us to get a working solution merged in so we can prevent breakages like the recent Verse breakage that was being wrangled today. Thanks for your help.

ceyhun committed:
Merge branch 'master' into rnmobile/initial_html_test

ceyhun committed:
Remove unneeded code after merging master

ceyhun committed:
Paste instead of typing when setting HTML content

ceyhun commented:
I think I found a way to paste HTML instead of typing. Let's see if this passes on CI and how long it takes. If it takes too long, we can check if we can still catch the crashes if we don't scroll at all. Maybe just rendering the block list is enough.

ceyhun committed:
Paste using keycode on Android

ceyhun committed:
Select html content view only on android

ceyhun committed:
Use iOS 14.0 simulator on server

ceyhun committed:
Print driver data on stop

ceyhun committed:
Enable remaining paragraph tests on iOS

ceyhun committed:
Remove dot at the end of jobURL

ceyhun committed:
Print menuButton

ceyhun committed:
Print error if environment cannot be initialized

ceyhun committed:
Revert "Use iOS 14.0 simulator on server"

This reverts commit 85178d15122d09bfb0060af2957ced110e41b41f.

ceyhun committed:
Revert "Print driver data on stop"

This reverts commit f136234b0d3cef5089162606a46b9191189447d8.

ceyhun committed:
Revert "Print menuButton"

This reverts commit 9e16a13eee9a52bc56528b31d3118c330a89aeac.

ceyhun committed:
Update server appium version to 1.18 and iOS to 14.0

ceyhun committed:
Add teardown to paste test

ceyhun committed:
On iOS wait for paste notification to disappear

ceyhun committed:
Double tap instead of long press to paste

ceyhun committed:
Fix lint error

ceyhun committed:
On iOS also long press before pasting

ceyhun committed:
On iOS use click instead of clickBeginningOfElement in paragraph test

ceyhun committed:
Scroll to bottom by adding a paragraph block to the end

jd-alexander commented:
@ceyhun I re-ran `End-to-End Tests / Admin - 1 (pull_request)` so it can pass.

jd-alexander commented:
@ceyhun out of curiosity I had a look at the video recordings of the test on Android since the build passed, and the content is not being pasted when the editor switches to HTML mode. On iOS, the content is pasted fine and the editor scrolls to the end 😄

ceyhun committed:
Use .type on Android instead of pressing paste keycode

ceyhun commented:
> @ceyhun out of curiosity I had a look at the video recordings of the test on Android since the build passed, and the content is not being pasted when the editor switches to HTML mode. On iOS, the content is pasted fine and the editor scrolls to the end 😄

@jd-alexander apparently pressing the paste keycode works on Sauce Labs but not on GitHub Actions. I'll try using the `.type()` method as before for Android instead, which seems to type instantly. Almost the same as paste, but instead of appending it replaces the content 😅 Thanks for checking the video recordings and catching this!

jd-alexander commented:
> @jd-alexander apparently pressing the paste keycode works on Sauce Labs but not on GitHub Actions. I'll try using the .type() method as before for Android instead, which seems to type instantly. Almost the same as paste, but instead of appending it replaces the content 😅 Thanks for checking the video recordings and catching this!

Of course. More than happy to help 🙇‍♂️

ceyhun commented:
@jd-alexander I think this is looking good now. Both GitHub Actions and Circle CI are passing. @geriux would you mind taking a look into this and gb-mobile PR please 🙏

jd-alexander commented:
> @jd-alexander I think this is looking good now. Both GitHub Actions and Circle CI are passing. @geriux would you mind taking a look into this and gb-mobile PR please 🙏

Thanks for these improvements @ceyhun The ios screen recording is spot on 👍 I have a question about the android screen recordings here. When I opened the file locally in QuickTime, I noticed that the video didn't show when the paragraph block was added to the canvas ( that would invoke the scroll to the last element behavior that ensures all blocks are rendered.) So I am checking if this is the expected video output or if I am seeing the incorrect video. Thank you. 

If it helps, I checked the Saucelab videos from the Full Test Suite for iOS and Android and the video output showed the last paragraph block that was added, and all blocks were rendered 😄 so I am wondering if this issue is related to Github Actions.

ceyhun commented:
> so I am wondering if this issue is related to Github Actions
It might be 🤔 I'll restart the job to see if it's any different.

ceyhun committed:
Merge branch 'master' into rnmobile/initial_html_test

ceyhun committed:
Check last block without scrolling on iOS

ceyhun committed:
Update package-lock.json

ceyhun commented:
@jd-alexander adding a paragraph block doesn't seem to scroll anymore (also on iOS). For Android I went back to scrolling via swiping but it still feels slow and will become slower as we add more blocks to the list. Do you think we might be able to speed it up even more? Maybe if we increase the swipe distance with each swipe?
For iOS the function seems to already return the block without scrolling so I added a similar function that only returns the element and doesn't scroll and used that instead.

jd-alexander commented:
> @jd-alexander adding a paragraph block doesn't seem to scroll anymore (also on iOS). For Android I went back to scrolling via swiping but it still feels slow and will become slower as we add more blocks to the list. Do you think we might be able to speed it up even more? Maybe if we increase the swipe distance with each swipe?

Good idea! If you like, I could take a look at modifying the coefficients used in the swipe functions to calculate the `endY` of the swipe operation to be a larger value, so that the swipe operation can move a lot faster. I know we have to strike an appropriate balance between the distance and time to ensure that rendering is taking place in a synchronized manner. 

> For iOS the function seems to already return the block without scrolling so I added a similar function that only returns the element and doesn't scroll and used that instead.

Sounds good!

jd-alexander committed:
added endYCoefficient to increase the swipe distance on Android.

jd-alexander committed:
Merge branch 'master' into rnmobile/initial_html_test

jd-alexander committed:
added package-lock.json

jd-alexander commented:
Hi @ceyhun so I reduced the time for the Android swipes and increased the swipe distance and within the Android logs for the E2E test, I saw a reduction of the time from 555.102s to 116.668s. I am working on verifying this some more. I noticed that the swiping behavior is still a bit erratic for some reason. 

1. When the swipe is initially triggered, no swipe takes place. I am not sure if this is because the editor hasn't rendered the content as yet, as the delay has been reduced so it's attempting swipes a lot quicker. 

2. Due to the nature of the algorithm in place to calculate the `endY` of the swipe operation, the height of the element being used cause the swipe to speed up at times, which I think is expected because we are assuming that if a control has a larger height then the swipe should be longer to get pass it. I am just noting that here. I think it's fine if we leave this as is,  but I am just noting it here.

Another question, for some reason I am getting a failure on the iOS test even though I didn't make any changes related to it. I was having some package-lock.json issues so I am wondering if syncing with master caused an issue. Let me know what you think. Thanks.

ceyhun commented:
> Another question, for some reason I am getting a failure on the iOS test even though I didn't make any changes related to it. I am hoping that the method that returns the last element on iOS without scrolling is not behaving in a flaky manner. The failed video looks similar to the success video. So I am not sure what how to interpret. I will work on reproducing this locally, I may need to downgrade my XCode because I was having some issues with it. Let me know what you think about the overall behavior. Thanks.

Thanks @jd-alexander 🙇 Appium version was updated on `master` branch, it may have broken this PR when we merged it. I couldn't see a video on GitHub Actions job and gb-mobile PR doesn't seem to have the latest changes. The error message on Actions is: `Unable to start WebDriverAgent session because of xcodebuild failure`. I'll try and see if I can reproduce locally and will attempt to fix.

ceyhun commented:
I couldn't reproduce the error locally using Xcode 12.2 and restarting the job also didn't resolve the issue ☹️

jd-alexander commented:
@ceyhun I am wondering if the issue is due to my update to the package-lock.json file. I have been having some issues with it and I was thinking to reinstall my entire npm, nvm setup because when I merge master into any of my branches, it creates a whole new package-lock.json file. So I thought I could solve this by just utilizing the package-lock.json that is on master, but I am seeing where you had done an update here 6b50987 so I am not sure if my sync messed things up in anyway. If necessary we could always revert to my commit before the merge to master and redo from there since that was the last successful build. I am analyzing the output from the logs for cues as well

```
Cannot initialize environment:  Error: [init({"browserName":"","platformName":"iOS","os":"iOS","deviceOrientation":"portrait","automationName":"XCUITest","appiumVersion":"1.18.3","app":"/Users/runner/work/gutenberg/gutenberg/packages/react-native-editor/ios/build/GutenbergDemo/Build/Products/Release-iphonesimulator/GutenbergDemo.app","processArguments":{"args":["uitesting"]},"deviceName":"iPhone 11"})] The environment you requested was unavailable.
An unknown server-side error occurred while processing the command. Original error: Unable to start WebDriverAgent session because of xcodebuild failure: An unknown server-side error occurred while processing the command. Original error: Could not proxy command to the remote server. Original error: connect ECONNREFUSED 127.0.0.1:8100
    at /Users/runner/work/gutenberg/gutenberg/node_modules/wd/lib/webdriver.js:154:15
    at Request._callback (/Users/runner/work/gutenberg/gutenberg/node_modules/wd/lib/http-utils.js:89:7)
    at Request.self.callback (/Users/runner/work/gutenberg/gutenberg/node_modules/request/request.js:185:22)
    at Request.emit (events.js:315:20)
    at Request.<anonymous> (/Users/runner/work/gutenberg/gutenberg/node_modules/request/request.js:1161:10)
    at Request.emit (events.js:315:20)
    at IncomingMessage.<anonymous> (/Users/runner/work/gutenberg/gutenberg/node_modules/request/request.js:1083:12)
    at Object.onceWrapper (events.js:421:28)
    at IncomingMessage.emit (events.js:327:22)
    at endReadableNT (internal/streams/readable.js:1327:12) {
  data: '{"status":13,"value":{"message":"An unknown server-side error occurred while processing the command. Original error: Unable to start WebDriverAgent session because of xcodebuild failure: An unknown server-side error occurred while processing the command. Original error: Could not proxy command to the remote server. Original error: connect ECONNREFUSED 127.0.0.1:8100"},"sessionId":null}'
```

ceyhun commented:
I think `package-lock.json` is in a correct state now. I didn't see any changes to it as well after running `npm install` locally, and also tests run fine locally. I believe this is the same flakiness that's happening on `master` right now. If you download the `appium-logs` [here](https://github.com/WordPress/gutenberg/actions/runs/581922910), you can see that it's the same error.

jd-alexander committed:
Merge branch 'trunk' into rnmobile/initial_html_test

jd-alexander committed:
Merge branch 'trunk' into rnmobile/initial_html_test

jd-alexander commented:
@ceyhun so I ran the `initial-html` block tests here locally on both Android and iOS and both tests succeeded. However, for some reason on Android, I see the full scroll effect - we are able to see the scroll go to the last block but on iOS, I don't see the scroll. Is this behavior okay or was it observed by you? 

 The issue I am observing is that the test runs on CI are a bit flaky. I was wondering if we can determine if it's safe to merge these changes since the flakiness is not due to the changes here but due to some underlying issue that is nullifying the `driver`? 
 
 We could do a couple more CI runs of this behavior once the video output and scroll behavior will allow us to verify that all components are appropriately rendered to detect breakages.

ceyhun commented:
> @ceyhun so I ran the initial-html block tests here locally on both Android and iOS and both tests succeeded. However, for some reason on Android, I see the full scroll effect - we are able to see the scroll go to the last block but on iOS, I don't see the scroll. Is this behavior okay or was it observed by you?

On iOS the `getLastElementByXPath` function seemed to return the block without scrolling, I'm guessing because of an eager loading of the block list, so in [this commit](https://github.com/WordPress/gutenberg/pull/26405/commits/cde770178eea3e4363c53706a43508b25962791e) I had added a similar function that only returns the element and doesn't scroll. We can scroll on iOS to the list's end as well just to be safe, in fact I think I tried [isDisplayed](http://appium.io/docs/en/commands/element/attributes/displayed/) to check if the last block on the list is displayed to stop scrolling, but `isDisplayed` didn't work as I expected it to. IIRC it was returning `false` although the element was "displayed". In the end I couldn't find another way to stop scrolling.

> I was wondering if we can determine if it's safe to merge these changes once we resolve this issue since the overall flakiness is not due to the changes here but due to some underlying issue that is nullifying the driver?

There was a recent attempt to fix the flakiness https://github.com/WordPress/gutenberg/pull/29589, we should probably get a merge and try again to see.

> We could do a couple more CI runs of this behavior once the video output and scroll behavior will allow us to verify that all components are appropriately rendered to detect breakages. The main concern I have is, I would want to be sure that these are as performant and reliable as possible since they will be run on every PR.

Couldn't agree more 💯

ceyhun committed:
Merge branch 'trunk' into rnmobile/initial_html_test

ceyhun commented:
> There was a recent attempt to fix the flakiness #29589, we should probably get a merge and try again to see.

Now that this PR is merged and `trunk` seems stable, let's give this another try!

jd-alexander commented:
> > There was a recent attempt to fix the flakiness #29589, we should probably get a merge and try again to see.
> 
> 
> 
> Now that this PR is merged and `trunk` seems stable, let's give this another try!

Thanks @ceyhun I was trying to update with trunk but I kept on getting some puppeteer errors saying the package is missing. Should I have just added it ?

ceyhun commented:
> Thanks @ceyhun I was trying to update with trunk but I kept on getting some puppeteer errors saying the package is missing. Should I have just added it ?

I didn't get any errors related to puppeteer 🤔 Just had a conflict in `packages/react-native-editor/jest_ui_test_environment.js`. 

BTW the Android one seems to have failed with an obscure error during the bundling step ☹️

jd-alexander commented:
> > Thanks @ceyhun I was trying to update with trunk but I kept on getting some puppeteer errors saying the package is missing. Should I have just added it ?
> 
> I didn't get any errors related to puppeteer 🤔 Just had a conflict in `packages/react-native-editor/jest_ui_test_environment.js`.

Okay. 🙇🏾 

> 
> BTW the Android one seems to have failed with an obscure error during the bundling step ☹️

Is there any way we can access those npm logs? I will do a re-run and see if it happens a second time.

jd-alexander commented:
@ceyhun I did a re-run and the test passed on Android. In the [video output](https://github.com/WordPress/gutenberg/suites/2281469241/artifacts/48072380) you will notice that the editor pauses for a bit and all the blocks are loaded and then the scroll operation begins and it scrolls to the end. The build before this failed with an ANR. 


https://user-images.githubusercontent.com/1509205/111737912-6e0b6800-884e-11eb-8673-fd6a0dcc567c.mp4


I have observed this in our Android test runs before. I  thought of disabling the System Isn't Respond dialog but IIRC, I had a discussion with @hypest about it and ANRs are an important signal that states something is wrong with the emulator environment that has caused resource constraints hence the test shouldn't run in that environment. 

Some ideas I can think of to somehow resolve this are : 

### 1. Catch ANR error output

 We could try to catch some error information about the ANR and report it. Can we dump content from the emulator to the logs/console in some way? If so we could leverage some ANR monitoring plugins that exist in the community. 
Bug snag has a component that they utilize in their [library](https://github.com/bugsnag/bugsnag-android/tree/master/bugsnag-plugin-android-anr) and then there's [ANR-WatchDog](https://github.com/SalomonBrys/ANR-WatchDog) that is a popular ANR detector used by four crash monitoring libraries.

Having the ANR output would help us in two ways.

 a) We could utilize the error output to understand what is causing the crash to take place. 
 b) We could restart jobs that fail due to the ANR since we can't determine if the emulator won't lock up on us sometimes. 

### 2. Change device profile of test emulator 

We could try changing the device profile to optimize the emulator. So we could increase the memory etc. to see if that will prevent these issues. The goal is to prevent ANRs and have predictable runs that won't be flaky. I have looked into the 
[android runner yml](https://github.com/WordPress/gutenberg/blob/1a52d1537e588264e92fce840a526d70004b57e1/.github/workflows/rnmobile-android-runner.yml) but I wasn't sure how I could make such modifications. Any clues here are appreciated. I will investigate how to configure properties. 

### 3. Disable ANR and let the test run. 

We could disable the ANR, though unsafe, I would say if we could disable it and the test runs then that should be fine. LinkedIn has a lib I have used in the past to do this when I was working on test suites with ANRs happening occasionally. This component [TestButler](https://github.com/linkedin/test-butler) disables the crash/ANRs, locks the cpu, wifi etc so that there aren't issues that can cause thread interruptions that make impact the test being run. I haven't investigated if this could be easily integrated with our current setup, as I wanted to get an opinion on the options here that should be explored first. 

Let me know if you have any thoughts @ceyhun @hypest about this. Thanks!

hypest commented:
> Let me know if you have any thoughts (@)ceyhun @hypest about this. Thanks!

Thanks for the analysis @jd-alexander ! Something not exactly clear to me is, is the ANR happening rather consistently in this PR but not on other tests already in place. If so, do we know why perhaps? I'd feel better if we fix the reason causing the ANR than sidestepping the ANR itself.

If the ANR happens in other tests that are already in place then it shouldn't be a blocker for this PR, unless this PR triggers the ANR more consistently.

jd-alexander commented:
> Thanks for the analysis @jd-alexander ! Something not exactly clear to me is, is the ANR happening rather consistently in this PR but not on other tests already in place. If so, do we know why perhaps? I'd feel better if we fix the reason causing the ANR than sidestepping the ANR itself.

🙇🏾 Good questions! 

> Is the ANR happening rather consistently in this PR but not on other tests already in place.

To investigate this, I am going to create a table in a comment below and do several re-runs to see if I can reproduce it consistently. 

> If the ANR happens in other tests that are already in place then it shouldn't be a blocker for this PR, unless this PR triggers the ANR more consistently.

Understood!

jd-alexander commented:
## React Native E2E Tests (Android & iOS ) / test (gutenberg-editor-blocks)- Test Runs

In this exercise, the Mobile E2E tests will be re-run several times to verify that the test behaves in a deterministic way.

| 🛠 Build 	| ⚙️ Android Status 	| ⚙️ iOS Status 	| 💬 Comments                           	|
|---------	|------------------	|--------------	|--------------------------------------	|
| 🛠 Run 1 	| 🟢 Success        	| 🟢 Success    	|                                      	|
| 🛠 Run 2 	| 🟢 Success        	| 🟢 Success    	|                                      	|
| 🛠 Run 3 	| 🔴 Failed          	| 🟢 Success    	| Appium Settings app did not respond. 	|
| 🛠 Run 4 	| 🟢 Success        	| 🟢 Success    	|                                      	|
| 🛠 Run 5 	| 🟢 Success        	| 🟢 Success    	|                                      	|
| 🛠 Run 6 	| 🟢 Success        	| 🟢 Success    	|                                      	|
| 🛠 Run 7 	| 🟢 Success        	| 🔴 Failed      	| lastBlockElement is undefined        	|
| 🛠 Run 8 	| 🟢 Success        	| 🟢 Success    	|                                      	|

------------
## Errors

### 🛠 Android - Run 3 - 🔴  Failed  
```
Setting RN platform to: android
Detected Android device running Android 9
E2E setup exception: Error: [init({"browserName":"","platformName":"Android","deviceName":"Android Emulator","automationName":"UiAutomator2","os":"Android","appPackage":"com.gutenberg","appActivity":"com.gutenberg.MainActivity","deviceOrientation":"portrait","appiumVersion":"1.18.1","app":"/Users/runner/work/gutenberg/gutenberg/packages/react-native-editor/android/app/build/outputs/apk/debug/app-debug.apk"})] The environment you requested was unavailable.
An unknown server-side error occurred while processing the command. Original error: Appium Settings app is not running after 5000ms
    at /Users/runner/work/gutenberg/gutenberg/node_modules/wd/lib/webdriver.js:154:15
    at Request._callback (/Users/runner/work/gutenberg/gutenberg/node_modules/wd/lib/http-utils.js:89:7)
    at Request.self.callback (/Users/runner/work/gutenberg/gutenberg/node_modules/request/request.js:185:22)
    at Request.emit (events.js:315:20)
    at Request.<anonymous> (/Users/runner/work/gutenberg/gutenberg/node_modules/request/request.js:1161:10)
    at Request.emit (events.js:315:20)
    at IncomingMessage.<anonymous> (/Users/runner/work/gutenberg/gutenberg/node_modules/request/request.js:1083:12)
    at Object.onceWrapper (events.js:421:28)
    at IncomingMessage.emit (events.js:327:22)
    at endReadableNT (internal/streams/readable.js:1327:12) {
  data: '{"status":13,"value":{"message":"An unknown server-side error occurred while processing the command. Original error: Appium Settings app is not running after 5000ms"},"sessionId":null}'
```

### 🛠 iOS - Run 7 - 🔴  Failed  

 expect(received).toBeTruthy()

    Received: undefined

      22 | 		}
      23 | 
    > 24 | 		expect( lastBlockElement ).toBeTruthy();
         | 		                           ^
      25 | 	} );
      26 | } );
      27 |

jd-alexander committed:
Merge branch 'trunk' into rnmobile/initial_html_test

jd-alexander commented:
@hypest Based on the results from the test runs it's safe to say that the ANR issues I noticed in the past are one off occurrences and we can proceed with the current implementation and setup of the android emulator runner 👍🏾 
If it ever reoccurs, I found an [issue](https://github.com/ReactiveCircus/android-emulator-runner/issues/129) in the [android-emulator-runner](https://github.com/ReactiveCircus/android-emulator-runner) repo where other members of the community are actively discussing solutions they have utilized to lessen ANRs.

@ceyhun on iOS, I was wondering if the `lastBlockElement undefined` error that occurs occasionally could be resolved with a timeout. In the video when the error occurs, all blocks are loaded so I was wondering if it may be time-related. I am not sure.

Apart from this, I think we are pretty close to getting this merged 👏🏾 #ProtectTheShip ⚓

ceyhun commented:
> Mobile E2E tests will be re-run several times to verify that the test behaves in a deterministic way.

That is at the same time awesome and also a huge amount of work to rerun those two again and again. Thank you so much @jd-alexander 🙇 

> @ceyhun on iOS, I was wondering if the lastBlockElement undefined error that occurs occasionally could be resolved with a timeout. In the video when the error occurs, all blocks are loaded so I was wondering if it may be time-related. I am not sure.

Do you know a way to access the previous run logs or the videos? I'm only able to access the latest one by clicking `Details` button next to the run and I also didn't get any emails for that failed test. I guess we can try adding a second call to `getLastElementByXPath` after some significant delay, if the first call isn't successful. Let me add this and also log the second calls to see how often it occurs.

ceyhun committed:
Retry one more time getting the last block after a delay on iOS

ceyhun committed:
Temporarily delete cancel workflow (and others) to run native jobs multiple times more quickly in consecutive commits

ceyhun committed:
Run 1

ceyhun committed:
Run 2

ceyhun committed:
Run 3

ceyhun committed:
Run 4

ceyhun committed:
Run 5

ceyhun committed:
Run 6

ceyhun committed:
Run 7

ceyhun committed:
Run 8

ceyhun commented:
@jd-alexander Pushed the changes including the delay. Now trying to run the native jobs consecutively without waiting by deleting other workflows, more specifically [the cancel workflow](https://github.com/WordPress/gutenberg/blob/trunk/.github/workflows/cancel.yml) (which cancels the previous jobs) and by pushing empty commits. I hope this works 😄 We should revert [this commit](https://github.com/WordPress/gutenberg/pull/26405/commits/fe5a1b11ed2f337eea40522f1e8620dd595e0aab), before merging.

jd-alexander commented:
> @jd-alexander Pushed the changes including the delay. Now trying to run the native jobs consecutively without waiting by deleting other workflows, more specifically [the cancel workflow](https://github.com/WordPress/gutenberg/blob/trunk/.github/workflows/cancel.yml) (which cancels the previous jobs) and by pushing empty commits. I hope this works 😄 We should revert [this commit](https://github.com/WordPress/gutenberg/pull/26405/commits/fe5a1b11ed2f337eea40522f1e8620dd595e0aab), before merging.

Thanks @ceyhun Such a nice trick to get consecutive re-runs 😉

jd-alexander commented:
> Do you know a way to access the previous run logs or the videos? I'm only able to access the latest one by clicking Details button next to the run and I also didn't get any emails for that failed test. I guess we can try adding a second call to getLastElementByXPath after some significant delay, if the first call isn't successful. Let me add this and also log the second calls to see how often it occurs.

When I try to access to the previous run logs and videos from my email, it points to the latest run so it seems we don't have access to the history which is odd.

ceyhun commented:
> Thanks @ceyhun Such a nice trick to get consecutive re-runs 😉

Seems like it didn't work 😅 I guess it's because there were conflicts, let me get another merge from trunk and only delete cancel job this time before trying again...

ceyhun committed:
Revert "Temporarily delete cancel workflow (and others) to run native jobs multiple times more quickly in consecutive commits"

This reverts commit fe5a1b11ed2f337eea40522f1e8620dd595e0aab.

ceyhun committed:
Merge branch 'trunk' into rnmobile/initial_html_test

ceyhun committed:
Temporarily delete cancel workflow

ceyhun committed:
Run 1

ceyhun committed:
Run 2

ceyhun committed:
Run 3

ceyhun committed:
Run 4

ceyhun committed:
Run 5

ceyhun committed:
Run 6

ceyhun committed:
Run 7

ceyhun committed:
Run 8

jd-alexander commented:
@ceyhun these are the log outputs of the runs. Do you have any thoughts on them?

### 🛠 Android - Run 2 - 🔴  Failed  

```
> @wordpress/react-native-editor@1.49.0 rn-bundle /Users/runner/work/gutenberg/gutenberg/packages/react-native-editor
> react-native bundle "--reset-cache" "--platform" "android" "--dev" "false" "--minify" "false" "--entry-file" "index.js" "--bundle-output" "android/app/src/main/assets/index.android.bundle" "--assets-dest" "android/app/src/main/res"

warning: the transform cache was reset.
emulator: ### WARNING: /etc/localtime does not point to zoneinfo-compatible timezone name

Assertion failed: (timeout != -1), function uv__io_poll, file ../deps/uv/src/unix/kqueue.c, line 256.
sh: line 1:  2880 Abort trap: 6           npm run rn-bundle -- --reset-cache --platform android --dev false --minify false --entry-file index.js --bundle-output android/app/src/main/assets/index.android.bundle --assets-dest android/app/src/main/res
npm ERR! code ELIFECYCLE
npm ERR! errno 134
npm ERR! @wordpress/react-native-editor@1.49.0 test:e2e:bundle:android: `mkdir -p android/app/src/main/assets && npm run rn-bundle -- --reset-cache --platform android --dev false --minify false --entry-file index.js --bundle-output android/app/src/main/assets/index.android.bundle --assets-dest android/app/src/main/res`
npm ERR! Exit status 134
npm ERR! 
npm ERR! Failed at the @wordpress/react-native-editor@1.49.0 test:e2e:bundle:android script.
```

### 🛠 Android - Run 4 - 🔴  Failed  

When I looked at the video output for this, the editor loads the content but it isn't able to scroll.
```
    [performTouchAction([{"action":"press","options":{"x":720,"y":797.3333333333334}},{"action":"wait","options":{"ms":100}},{"action":"moveTo","options":{"x":720,"y":0}},{"action":"release","options":{}}])] Error response status: 12, , InvalidElementState - An element command could not be completed because the element is in an invalid state (e.g. attempting to click a disabled element). Selenium error: Swipe action cannot be performed

      at exports.newError (../../node_modules/wd/lib/utils.js:152:13)
      at ../../node_modules/wd/lib/callbacks.js:94:19
      at ../../node_modules/wd/lib/webdriver.js:202:5
      at Request._callback (../../node_modules/wd/lib/http-utils.js:89:7)
      at Request.self.callback (../../node_modules/request/request.js:185:22)
      at Request.<anonymous> (../../node_modules/request/request.js:1161:10)
      at IncomingMessage.<anonymous> (../../node_modules/request/request.js:1083:12)

Test Suites: 1 failed, 1 total
```


### Potential solutions so far

A common theme I noticed is that sometimes the emulator takes a while to be booted. The log output below is included in both logs. I am not sure if this should be normal or not. 

```
The process '/Users/runner/Library/Android/sdk/platform-tools/adb' failed with exit code 1
/Users/runner/Library/Android/sdk/platform-tools/adb shell getprop sys.boot_completed
adb: device offline
The process '/Users/runner/Library/Android/sdk/platform-tools/adb' failed with exit code 1
/Users/runner/Library/Android/sdk/platform-tools/adb shell getprop sys.boot_completed
adb: device offline
The process '/Users/runner/Library/Android/sdk/platform-tools/adb' failed with exit code 1
/Users/runner/Library/Android/sdk/platform-tools/adb shell getprop sys.boot_completed
The process '/Users/runner/Library/Android/sdk/platform-tools/adb' failed with exit code 1
adb: device offline
```

Currently, the `android-emulator-runner` isn't the latest version. It is currently 2.14.3
https://github.com/WordPress/gutenberg/blob/a49bd8a7a250866e04805bfc11534c3f1f4be96d/.github/workflows/rnmobile-android-runner.yml#L38
 and there's a newer release, [2.15.0 ](https://github.com/ReactiveCircus/android-emulator-runner/releases/tag/v2.15.0)that contains some important changes that:

1. Set the emulators default `hw.cpu.ncore` to 2. 
2. We can now add the `cores` config to the workflow to modify the emulator cores.

What do you think about upgrading to `v2.15.0` and doing several more runs to see how the emulator performs?

ceyhun commented:
Thanks a lot for checking the runs @jd-alexander 🙇 I took a quick look to the iOS ones and the retry of the `getLastElementByXPath` after a delay was happening in a bunch of them, but fortunately all of them passed after the retry 🎉 

> @ceyhun these are the log outputs of the runs. Do you have any thoughts on them?

`
Assertion failed: (timeout != -1), function uv__io_poll, file ../deps/uv/src/unix/kqueue.c, line 256.
sh: line 1:  2880 Abort trap: 6           npm run rn-bundle -- --reset-cache --platform android --dev false --minify false --entry-file index.js --bundle-output android/app/src/main/assets/index.android.bundle --assets-dest android/app/src/main/res
`

This one seems to me like the machine was out of memory during the JS bundle step. 


`
Error response status: 12, , InvalidElementState - An element command could not be completed because the element is in an invalid state (e.g. attempting to click a disabled element). Selenium error: Swipe action cannot be performed
`

Not sure about this one though. Could it be that the emulator just froze?

> What do you think about upgrading to v2.15.0 and doing several more runs to see how the emulator performs?

If the memory error keeps happening, we can try to split the steps (bundle JS, build APK, run tests) that are currently done within a single script `test:e2e:android:local` and maybe bundle JS before launching the emulator. But maybe it's better to try upgrading to v2.15.0 before going that way.

jd-alexander commented:
> Thanks a lot for checking the runs @jd-alexander 🙇 I took a quick look to the iOS ones and the retry of the getLastElementByXPath after a delay was happening in a bunch of them, but fortunately all of them passed after the retry 🎉

Awesome! 

> Not sure about this one though. Could it be that the emulator just froze?

I think that's what happened. 

>  But maybe it's better to try upgrading to v2.15.0 before going that way.

I agree. Would you mind doing the upgrade and triggering 8 more runs to see what happens? I would do it but I am not sure which approach to utilize. I see where a hash is being used so I am not sure if that's supposed to remain. Also the android emulator runner maintainers said going forward folks should use `v2` version to include all updates but that would mean any new releases would be pulled in automatically without testing so including the minor version ( `v2.15`) is best right?

ceyhun commented:
> I agree. Would you mind doing the upgrade and triggering 8 more runs to see what happens? I would do it but I am not sure which approach to utilize. I see where a hash is being used so I am not sure if that's supposed to remain. Also the android emulator runner maintainers said going forward folks should use `v2` version to include all updates but that would mean any new releases would be pulled in automatically without testing so including the minor version ( `v2.15`) is best right?

Sure 👍 I think they were introduced [here](https://github.com/WordPress/gutenberg/pull/29485) and are for security purposes and maybe we should keep using a hash.

ceyhun committed:
Merge branch 'trunk' into rnmobile/initial_html_test

ceyhun committed:
Temporarily delete non-native workflows

ceyhun committed:
Bump reactivecircus/android-emulator-runner action to v2.15.0

ceyhun committed:
Run 1

ceyhun committed:
Run 2

ceyhun committed:
Run 3

ceyhun committed:
Run 4

ceyhun committed:
Run 5

ceyhun committed:
Run 6

ceyhun committed:
Run 7

ceyhun committed:
Run 8

ceyhun commented:
Bumped the action hash to use v2.15.0, and also tried temporarily deleting the non-native jobs again. We should revert these 2 commits before merging: 
- b31695a7bb55efb19b2740ac72cdbc3dd69a1a18
- ff5a3c4b8ade129c3057868b2bb47b60522f54fe

ceyhun committed:
Run 9

ceyhun committed:
Run 10

ceyhun committed:
Run 11

ceyhun committed:
Run 12

ceyhun committed:
Run 13

ceyhun committed:
Run 14

ceyhun committed:
Run 15

ceyhun committed:
Run 16

ceyhun commented:
Yay 🎉🎊🎆 All the 8 are green ✅ Let's run 8 more!!!

BTW here's the command I use to push empty commits:
```sh
for i in `seq 1 8`; do git commit --allow-empty -m "Run $i" && git push; done
```

jd-alexander commented:
> Yay 🎉🎊🎆 All the 8 are green ✅ Let's run 8 more!!!

Awesome!!!! 🥳

> Bumped the action hash to use v2.15.0, and also tried temporarily deleting the non-native jobs again. We should revert these 2 commits before merging:
b31695a
ff5a3c4

I will keep the reverting of these commits in mind. I am hoping the v2.15.0 really stabilises things on Android. Let's see what these 8 new runs have in store.

> 
> BTW here's the command I use to push empty commits:
> 
> ```shell
> for i in `seq 1 8`; do git commit --allow-empty -m "Run $i" && git push; done
> ```

Pretty neat. Beats doing it manually 😅

jd-alexander commented:
The results are in! We got 16 consecutive ✅ successful runs @ceyhun Things are looking good 😎🚀🚇🚢

jd-alexander commented:
@ceyhun I just saw the conversation about flaky tests [here](https://github.com/WordPress/gutenberg/pull/30344#issuecomment-809645187) 👍🏾  👍🏾  👍🏾  👍🏾 

I am quite confident we are in a good place from a stability standpoint. When we do merge these changes we can monitor for stability and if any hiccups arise we have several Android-related desired capabilities ( in particular timeout-related caps ) that we can leverage to resolve any flaky issues that may occur.

ceyhun committed:
Merge branch 'trunk' into rnmobile/initial_html_test

ceyhun committed:
Revert "Temporarily delete cancel workflow"

This reverts commit b31695a7bb55efb19b2740ac72cdbc3dd69a1a18.

ceyhun committed:
Revert "Temporarily delete non-native workflows"

This reverts commit ff5a3c4b8ade129c3057868b2bb47b60522f54fe.

ceyhun commented:
@jd-alexander I've reverted the commits deleting other workflows and also updated the gb-mobile PR with the gutenberg's current ref. If that's also green, I think we can go forward with both 🎉

ceyhun commented on .github/workflows/rnmobile-android-runner.yml:
```diff
@@ -11,7 +11,7 @@ jobs:
       matrix:
         native-test-name: [
-          gutenberg-editor-gallery
+          gutenberg-editor-blocks
```
What do you think about renaming this to `gutenberg-editor-initial-html` and also keeping `gutenberg-editor-gallery`, but maybe making them run one after another instead of running in parallel?

jd-alexander replied:
I agree with the renaming. What is the rationale for keeping the gallery test ? I thought this test would cover all cases and I'm wondering if that test could be more flaky than this one. Also I'm wondering about the collective time of both them running one after the other as well. Let me know what you think.

ceyhun replied:
To be honest, I'm not interested in gallery block in this case 😄 But the gallery test checks some other aspects of the editor that are not covered by the initial HTML tests, mainly around the inserter menu, like opening the inserter menu, scrolling the inserter menu, and inserting a block.

The primary time-consuming task is building the js bundle and the native code (the ipa and the apk). The gallery test itself runs in ~5 seconds on iOS and ~50 seconds on Android. Android is slower due to iOS having a larger inserter, so there's no need to scroll the inserter to find the gallery block on iOS. 

Gallery test has been around for quite some time now, and I think it should be stable enough. But I realized now that we had made some changes around scrolling in this PR, that could be affecting the gallery test's stability. So it could be better to leave the gallery test out for now and maybe add it back in, in a separate PR making sure again it's flawless 💯

jd-alexander replied:
> To be honest, I'm not interested in gallery block in this case 😄 But the gallery test checks some other aspects of the editor that are not covered by the initial HTML tests, mainly around the inserter menu, like opening the inserter menu, scrolling the inserter menu, and inserting a block.

I understand. This makes a lot of sense. 

> The primary time-consuming task is building the js bundle and the native code (the ipa and the apk). The gallery test itself runs in ~5 seconds on iOS and ~50 seconds on Android. Android is slower due to iOS having a larger inserter, so there's no need to scroll the inserter to find the gallery block on iOS
> Gallery test has been around for quite some time now, and I think it should be stable enough. But I realized now that we had made some changes around scrolling in this PR, that could be affecting the gallery test's stability. So it could be better to leave the gallery test out for now and maybe add it back in, in a separate PR making sure again it's flawless 💯

Sounds like a good plan to me to re-add it after we have done some more tests! Thanks for expounding here @ceyhun

jd-alexander commented:
> @jd-alexander I've reverted the commits deleting other workflows and also updated the gb-mobile PR with the gutenberg's current ref. If that's also green, I think we can go forward with both 🎉

Agreed. Thanks @ceyhun 😄 Once we have resolved the comment, we are good to go.

ceyhun committed:
Merge branch 'trunk' into rnmobile/initial_html_test

ceyhun committed:
Rename test file
